src/BenchmarkDotNet/ConsoleArguments/ConfigParser.cs

```diff
@@ -410,11 +408,21 @@ private static Job CreateJobForGivenRuntime(Job baseJob, string runtimeId, Comma
                     return MakeMonoAOTLLVMJob(baseJob, options, "net6.0");
                 case RuntimeMoniker.MonoAOTLLVMNet70:
                     return MakeMonoAOTLLVMJob(baseJob, options, "net7.0");
+                case RuntimeMoniker.Mono70:
+                    return CreateCoreJob(baseJob, "net7.0", "Mono with .NET 7.0", options, CoreRuntime.Core70)
+                        .WithArguments(new Argument[] { new MsBuildArgument("/p:UseMonoRuntime=true") });
                 default:
                     throw new NotSupportedException($"Runtime {runtimeId} is not supported");
             }
         }
 
+        private static Job CreateCoreJob(Job baseJob, string runtimeId, string runtimeName, CommandLineOptions options, Runtime runtime)
+        {
+            return baseJob
+                .WithRuntime(runtime)
+                .WithToolchain(CsProjCoreToolchain.From(new NetCoreAppSettings(runtimeId, null, runtimeName, options.CliPath?.FullName, options.RestorePath?.FullName)));

```

Comment:
`CsProjCoreToolchain` uses `DotNetCliBuilder` for building:

https://github.com/dotnet/BenchmarkDotNet/blob/1f5637c784ad9887a2d5abdba54d1337655281b1/src/BenchmarkDotNet/Toolchains/CsProj/CsProjCoreToolchain.cs#L38

which performs `dotnet build` to build the benchmark:

https://github.com/dotnet/BenchmarkDotNet/blob/1f5637c784ad9887a2d5abdba54d1337655281b1/src/BenchmarkDotNet/Toolchains/DotNetCli/DotNetCliBuilder.cs#L39

and then `DotNetCliExecutor` to run the executable:

https://github.com/dotnet/BenchmarkDotNet/blob/1f5637c784ad9887a2d5abdba54d1337655281b1/src/BenchmarkDotNet/Toolchains/CsProj/CsProjCoreToolchain.cs#L39

the executor basically performs `dotnet $pathToLibrary.dll` to run the benchmarks:

https://github.com/dotnet/BenchmarkDotNet/blob/1f5637c784ad9887a2d5abdba54d1337655281b1/src/BenchmarkDotNet/Toolchains/DotNetCli/DotNetCliExecutor.cs#L72

You most likely need to define a new toolchain that consists of:
* `CsProjGenerator` same way as `CsProjCoreToolchain` does: https://github.com/dotnet/BenchmarkDotNet/blob/1f5637c784ad9887a2d5abdba54d1337655281b1/src/BenchmarkDotNet/Toolchains/CsProj/CsProjCoreToolchain.cs#L37
* `DotNetCliPublisher` with `extraArguments` set to `-r $runtimeId`: https://github.com/dotnet/BenchmarkDotNet/blob/1f5637c784ad9887a2d5abdba54d1337655281b1/src/BenchmarkDotNet/Toolchains/DotNetCli/DotNetCliPublisher.cs#L11
You can take a look at `NatvieAOTToolchain` to see how it provides the runtime id: https://github.com/dotnet/BenchmarkDotNet/blob/1f5637c784ad9887a2d5abdba54d1337655281b1/src/BenchmarkDotNet/Toolchains/NativeAot/NativeAotToolchain.cs#L46
* `Executor` that runs provided `.exe` rather than `dotnet $pathTo.dll`: https://github.com/dotnet/BenchmarkDotNet/blob/1f5637c784ad9887a2d5abdba54d1337655281b1/src/BenchmarkDotNet/Toolchains/Executor.cs

The tricky part will be to make sure that `ExecutablePath` provided by `CsProjGenerator` is the one that points to the executable build by `dotnet publish`, not the `dll` build by `dotnet build`.

It's used here by the executor:

https://github.com/dotnet/BenchmarkDotNet/blob/1f5637c784ad9887a2d5abdba54d1337655281b1/src/BenchmarkDotNet/Toolchains/Executor.cs#L27

It comes from here:

https://github.com/dotnet/BenchmarkDotNet/blob/1f5637c784ad9887a2d5abdba54d1337655281b1/src/BenchmarkDotNet/Toolchains/GeneratorBase.cs#L114

https://github.com/dotnet/BenchmarkDotNet/blob/1f5637c784ad9887a2d5abdba54d1337655281b1/src/BenchmarkDotNet/Toolchains/CsProj/CsProjGenerator.cs#L52

You might need to introduce a new type that derives from `CsProjGenerator` and does this:

https://github.com/dotnet/BenchmarkDotNet/blob/1f5637c784ad9887a2d5abdba54d1337655281b1/src/BenchmarkDotNet/Toolchains/NativeAot/Generator.cs#L66-L67

Reply:
Thanks for very well-detailed explanation. Everything is clear, just few notes you might want to comment on:
> DotNetCliPublisher with extraArguments set to -r $runtimeId

I will do this because I see there is a way to obtain it automatically anyway (so that it is still user-friendly). But your `small repro` returns `true`  with `dotnet publish -c Release -f net6.0 --self-contained -r win-x64 /p:UseMonoRuntime=true` as well as with `dotnet publish -c Release -f net6.0 --self-contained /p:UseMonoRuntime=true` on my local Windows 10.

> Executor that runs provided .exe rather than dotnet $pathTo.dll:

Same thing here. Both `.exe` execution and `dotnet $pathTo.dll` works fine locally.
However, I am not sure what happens on Linux and Mac. I can check on Linux using Docker or WSL if you are interested in, otherwise I will just do what you suggested (just ignore these notes then).

Reply:
> dotnet publish -c Release -f net6.0 --self-contained /p:UseMonoRuntime=true

This is something new to me, I remember dotnet SDK always complaining about lack of runtime id when running `--self-contained`. But the produced exe is put to the runtime-specific folder, so we might need it anyway:

```log
PS D:\projects\repros\consoleBdn> dotnet publish -c Release -f net7.0 --self-contained /p:UseMonoRuntime=true
MSBuild version 17.4.0-preview-22428-01+14c24b2d3 for .NET
  Determining projects to restore...
  All projects are up-to-date for restore.
C:\Program Files\dotnet\sdk\7.0.100-rc.1.22431.12\Sdks\Microsoft.NET.Sdk\targets\Microsoft.NET.RuntimeIdentifierInference.targets(219,5): message NETSDK1057: You are using a preview version of .NET. See: https://aka.ms/dotnet-support-p
olicy [D:\projects\repros\consoleBdn\consoleBdn.csproj]
  consoleBdn -> D:\projects\repros\consoleBdn\bin\Release\net7.0\win-x64\consoleBdn.dll
  consoleBdn -> D:\projects\repros\consoleBdn\bin\Release\net7.0\win-x64\publish\
```

> However, I am not sure what happens on Linux and Mac

If you add a test similar to what I've described in https://github.com/dotnet/BenchmarkDotNet/pull/2142#discussion_r992741339 and it works on your Windows machine it should work on Linux and Mac too and it's enough to just use CI to verify that.

Reply:
Thank you @adamsitnik, this is done. Could you please take a look?